File: app/src/main/java/com/asmat/rolando/popularmovies/adapters/CastLinearAdapter.java
package com.asmat.rolando.popularmovies.adapters;

import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.asmat.rolando.popularmovies.R;
import com.asmat.rolando.popularmovies.models.AdapterOnClickHandler;
import com.asmat.rolando.popularmovies.models.Cast;
import com.asmat.rolando.popularmovies.models.RoundedTransformation;
import com.squareup.picasso.Picasso;

public class CastLinearAdapter extends BaseLinearAdapter<Cast, CastLinearAdapter.ViewHolder> {

    public CastLinearAdapter(AdapterOnClickHandler<Cast> clickHandler) {
        super(clickHandler);
    }

    @Override
    int getLayoutForLinearItem() {
        return R.layout.cast_linear_item;
    }

    @Override
    ViewHolder createViewHolder(View view) {
        return new CastLinearAdapter.ViewHolder(view);
    }

    class ViewHolder extends BaseLinearAdapter.ViewHolder {
        private final ImageView thumbnail;
        private final TextView name;
        private final TextView role;

        ViewHolder(View itemView) {
            super(itemView);
            thumbnail = itemView.findViewById(R.id.thumbnail);
            name = itemView.findViewById(R.id.name);
            role = itemView.findViewById(R.id.role);
            thumbnail.setOnClickListener(this);
        }

        @Override
        void bind(Object item) {
            if (item instanceof Cast) {
                Cast cast = (Cast) item;
                name.setText(cast.getName());
                role.setText(cast.getCharacter());
                String url = cast.getThumbnailURL();
                Picasso.with(thumbnail.getContext())
                        .load(url)
                        .resize(342, 513)
                        .centerCrop()
                        .transform(new RoundedTransformation(50, 0))
                        .error(R.drawable.person)
                        .into(thumbnail);
            }
        }
    }
}

File: app/src/main/java/com/asmat/rolando/popularmovies/ui/adapters/linear/CastLinearAdapter.kt
package com.asmat.rolando.popularmovies.ui.adapters.linear

import android.view.View
import android.widget.ImageView
import android.widget.TextView

import com.asmat.rolando.popularmovies.R
import com.asmat.rolando.popularmovies.networking.the.movie.db.models.CreditsResponse
import com.asmat.rolando.popularmovies.ui.transformations.RoundedTransformation
import com.asmat.rolando.popularmovies.utilities.URLUtils
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.cast_linear_item.view.*

private typealias Cast = CreditsResponse.Cast
private typealias CastViewHolder = CastLinearAdapter.ViewHolder

class CastLinearAdapter(clickHandler: (Cast) -> Unit?) :
        BaseLinearAdapter<Cast, CastViewHolder>(clickHandler) {

    override val layoutForLinearItem: Int
        get() = R.layout.cast_linear_item

    override fun createViewHolder(view: View): ViewHolder {
        return ViewHolder(view)
    }

    inner class ViewHolder(view: View) : BaseLinearAdapter<Cast, CastViewHolder>.ViewHolder(view) {

        private val thumbnail: ImageView = view.thumbnail
        private val name: TextView = view.name
        private val role: TextView = view.role

        init {
            thumbnail.setOnClickListener(this)
        }

        override fun bind(item: Cast) {
            name.text = item.name
            role.text = item.character
            item.profile_path?.let { profilePath ->
                val url = URLUtils.getImageURL342(profilePath)
                Picasso.with(thumbnail.context)
                        .load(url)
                        .resize(342, 513)
                        .centerCrop()
                        .transform(RoundedTransformation(50, 0))
                        .error(R.drawable.person)
                        .into(thumbnail)
            }
        }
    }
}
